### PR TITLE
fix(chart-data-api): assert referenced columns are present in datasource

### LINKED ISSUE #10411: form_data on request could be editable, and run a custom query
The form_data send on a request, could be modified by a "gamma" user and read all the content from the database.
By modifying the request, we are able to forge our own query and interact with the current database in a manner not provided by the application. For instance, we can get, the current database, the current user, the database version.
By modifying the "datasource" parameter (which is in the format XX_table, where XX is the ID, we can list all the tables connected with superset.

### Screenshots
Getting all the tables by changing the ID on XX_table:
image

Finding the database and version:
image

Changing tables and values to add on the query:
image

- superset version: `0.35`
- python version: `3.6`

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
Currently it might be possible to perform limited SQL injection via both the legacy and new chart data API by manipulating the `columns`, `groupby`, `filters` or `metrics` properties of `QueryObject`. This plugs that hole both on the new and legacy chart data API.

Further work is required to ensure other fields aren't vulnerable to SQL injection. However, to keep the PR size manageable I suggest limiting the scope of this PR to fields with direct column references and adding more assertions later to the `where` and `having` fields + `SQL` type ad-hoc metrics.

### TEST PLAN
- CI
- new tests
- manual testing - clicked through all examples dashboards and didn't notice any chart breakage

### ADDITIONAL INFORMATION

- [x] Has associated issue: closes #10411
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by willbarrett — APPROVED** [review]
LGTM once CI passes

**Comment by villebro** [comment]
@willbarrett I'll add a few more assertions and improve the tests, should be done in a few days.

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -90,6 +90,19 @@ class AnnotationDatasource(BaseDatasource):
...
+        "id",
+        "start_dttm",
+        "end_dttm",
+        "layer_id",
+        "short_descr",
+        "long_descr",
+        "json_metadata",
+        "created_by_fk",
+        "changed_by_fk",
+    ]
```
Comment (villebro) [inline superset/connectors/sqla/models.py:105]: These had to be added, as `AnnotationDatasource`s don't have any defined `columns`.

File: tests/query_context_tests.py [hunk 2]
```diff
@@ -172,13 +173,58 @@ def test_csv_response_format(self):
...
+        Ensure that calling invalid columns names in groupby are caught
+        """
+        self.login(username="admin")
+        table_name = "birth_names"
+        table = self.get_table_by_name(table_name)
+        payload = get_query_context(table.name, table.id, table.type)
+        payload["queries"][0]["groupby"] = ["currentDatabase()"]
+        query_context = ChartDataQueryContextSchema().load(payload)
+        query_payload = query_context.get_payload()
+        assert query_payload[0].get("error") is not None
```
Comment (dpgaspar) [inline tests/query_context_tests.py:194]: is `assert` preferable to `self.assertEqual` or `self.assertIsNotNone` ?

File: tests/query_context_tests.py [hunk 3]
```diff
@@ -172,13 +173,58 @@ def test_csv_response_format(self):
...
+    def test_sql_injection_via_filters(self):
+        """
+        Ensure that calling invalid columns names in filters are caught
+        """
+        self.login(username="admin")
+        table_name = "birth_names"
+        table = self.get_table_by_name(table_name)
+        payload = get_query_context(table.name, table.id, table.type)
+        payload["queries"][0]["groupby"] = ["name"]
+        payload["queries"][0]["metrics"] = []
```
Comment (dpgaspar) [inline tests/query_context_tests.py:220]: Are we able to deny an injected `metric`?

File: tests/query_context_tests.py [hunk 4]
```diff
@@ -172,13 +173,58 @@ def test_csv_response_format(self):
...
 
+    def test_sql_injection_via_groupby(self):
+        """
+        Ensure that calling invalid columns names in groupby are caught
+        """
+        self.login(username="admin")
+        table_name = "birth_names"
+        table = self.get_table_by_name(table_name)
+        payload = get_query_context(table.name, table.id, table.type)
+        payload["queries"][0]["groupby"] = ["currentDatabase()"]
```
Comment (dpgaspar) [inline tests/query_context_tests.py:191]: If `currentDatabase()` is a defined metric will it run ok?

File: tests/query_context_tests.py (hunk 2 above)
Comment (villebro) [reply tests/query_context_tests.py:194]: Yes, this is the `pytest` way

**Review by dpgaspar — COMMENTED** [review]
Some doubts regarding how metrics are handled

File: tests/query_context_tests.py (hunk 4 above)
Comment (villebro) [reply tests/query_context_tests.py:191]: It shouldn't, as it isn't an aggregate expression, hence will be missing from the groupby causing an invalid query.

File: tests/query_context_tests.py (hunk 3 above)
Comment (villebro) [reply tests/query_context_tests.py:220]: I haven't been able to create one yet, would be interested to see if someone is able to do one.

Comment (dpgaspar) [reply tests/query_context_tests.py:220]: Not easy that's for sure

**Review by willbarrett — APPROVED** [review]

**Pull request merged**